### ISSUE TITLE
nodemail 'Module not found' with EmailProvider

### ISSUE THREAD

**Issue by johanguse** [nextauthjs/next-auth#10501]
### Provider type

Email

### Environment

  System:
    OS: macOS 14.4.1
    CPU: (8) arm64 Apple M1
    Memory: 403.78 MB / 16.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 21.6.1 - ~/.nvm/versions/node/v21.6.1/bin/node
    Yarn: 1.22.21 - /opt/homebrew/bin/yarn
    npm: 10.2.4 - ~/.nvm/versions/node/v21.6.1/bin/npm
    pnpm: 8.15.6 - ~/Library/pnpm/pnpm
    bun: 1.1.1 - /opt/homebrew/bin/bun
  Browsers:
    Brave Browser: 123.1.64.113
    Chrome: 123.0.6312.107
    Chrome Canary: 125.0.6405.0
    Edge: 123.0.2420.81
    Firefox Nightly: 96.0a1
    Safari: 17.4.1
  npmPackages:
    @auth/prisma-adapter: ^1.5.2 => 1.5.2 
    next: 14.1.4 => 14.1.4 
    next-auth: beta => 5.0.0-beta.16 
    react: ^18.2.0 => 18.2.0 

### Reproduction URL

https://github.com/mickasmt/next-saas-stripe-starter

### Describe the issue

I don't know why I got some errors on nodemail module using EmailProvider

More people have the same problem here: https://github.com/nextauthjs/next-auth/discussions/7983

Types of error I got:

Module not found: Can't resolve 'stream'
Module not found: Can't resolve 'child_process'
Module not found: Can't resolve 'http'

I got same type of error with this two imports
`import EmailProvider from "next-auth/providers/email"
`
`import EmailProvider from "next-auth/providers/nodemail"
`

![image](https://github.com/nextauthjs/next-auth/assets/6184866/cf6c0b85-06b6-485a-8ad8-b47f9fcd0786)

This is weird because this guy here is doing the same job with EmailProvider but it's works. Maybe it's only works on server side?

https://github.com/pjborowiecki/SAASY-LAND-Next-14-Starters-With-Authentication-And-Database-Implemented/blob/main/src/config/auth.ts

### How to reproduce

Uncomment the lines related to EmailProvider on auth.config.ts and try to run the project

https://github.com/mickasmt/next-saas-stripe-starter/blob/main/auth.config.ts#L16

### Expected behavior

The project should run fine, current I can run or build my NextJS project

**Comment by balazsorban44** [comment]
Nodemailer - as the name suggests - only works in Node.js environemnts, not in the browser or the Edge runtime.

We have a Resend-specific provider, which doesn't use nodemalier at all, since: https://github.com/nextauthjs/next-auth/pull/9753. This should greatly simplify your setup!

**Comment by willemmulder** [comment]
I have the same issue, but cannot switch to another Provider.

@balazsorban44 Does the nextauthjs middleware actually require the full authOptions, or can we do with e.g. just access to the Adapter, and not the Providers? That would solve this problem, because the Providers cause the edge issues, and they might not even be used by the middleware?

**Comment by johanguse** [comment]
Use Resend @willemmulder on `"next-auth": "beta",`

```
import Resend from 'next-auth/providers/resend'

Resend({
      id: 'resend',
      name: 'Resend',
      apiKey: env.RESEND_API_KEY,
      from: env.RESEND_FROM_EMAIL,
})
```

**Comment by willemmulder** [comment]
@johanguse That won't work for us. We send email through the Gmail SMTP. But that's not too relevant - there could be many reasons why people want to use nodemailer or have dependencies on other libraries :-) 

Do you know which part of the authOptions object is actually used by the middleware / auth function?

**Comment by osdiab** [comment]
Agreed with @willemmulder , I realized that because we use `jsx-email` when configuring the 
email provider, as well as set cookies on nextJS server side in our auth callbacks, we cannot have our middleware use the `auth()` function despite that logic being basically irrelevant for the middleware to consume.

It seems like there ought to be some way to use the `auth()` function in middleware, without needing to pull in the details of how the adapter works, or how emails are sent, as the only reason I want to get it is to get basic session data - specifically the logged in user's ID.

**Comment by MoOx** [comment]
This is somehow confusing. Doc encourage you to use a single `auth()` call to [replace various ones like `useSession()`](https://authjs.dev/getting-started/migrating-to-v5#authentication-methods) but when you do that, you have problem like frontend trying to load code for backend.

How can we use the same function if architecture is not prepared for this ?

**Comment by MoOx** [comment]
@balazsorban44 you cannot just say to people "hey use this service" when we all have constraint like our boss saying "we are going to send mail using Xxxx". How can you consider this not a problem ?

**Comment by balazsorban44** [comment]
We recently wrote an article on Edge compatibility. Check it out: https://authjs.dev/guides/edge-compatibility

@MoOx I simply recommended Resend, since the OP's reproduction suggested they were using that.

I personally don't care what you or your boss wants to use at your company.

However if you, your boss and/or company uses my/our open source project and earn money from it, you can rather consider sponsoring it.

_And please, don't be like [me](bal.so/early-oss)_

**Comment by balazsorban44** [comment]
We recently wrote an article on Edge compatibility. Check it out: https://authjs.dev/guides/edge-compatibility

@MoOx I simply recommended Resend, since the OP's reproduction suggested they were using that.

I personally don't care what you or your boss wants to use at your company.

However if you, your boss and/or company uses my/our open source project and earn money from it, you can rather consider sponsoring it.

_And please, don't be like [me](bal.so/early-oss)_

**Comment by balazsorban44** [comment]
We recently wrote an article on Edge compatibility. Check it out: https://authjs.dev/guides/edge-compatibility

@MoOx I simply recommended Resend, since the OP's reproduction suggested they were using that.

I personally don't care what you or your boss wants to use at your company.

However if you, your boss and/or company uses my/our open source project and earn money from it, you can rather consider sponsoring it.

_And please, don't be like [me](bal.so/early-oss)_

**Comment by MoOx** [comment]
Sorry if my message appeared as rude. I wasn't my intention. I know how maintaining OSS can be.

The thing that surprised me is that configuring provider for email involve `import`ing the code into frontend bundle, which seems weird to me as we (as dev/user) could expect this to be handled for a easy/amazing bundle as next-auth appear to be from a quick look.
Not sure if I am clear but I guess that nodemailer isn't really required for next-auth frontend runtime, so I was weirdly surprised that out of the box, it's not working without trouble and/or weird configuration for webpack to ignore package fs etc.

**Comment by balazsorban44** [comment]
> The thing that surprised me is that configuring provider for email involve importing the code into frontend bundle

This is false though. Where did you see this happening? Sounds like a misconfig.

**Comment by osdiab** [comment]
@MoOx I believe it doesn't get into your frontend generally since you wouldn't directly use the `auth()` value in your frontend, in the case of React you'd trigger logins etc with `next-auth/react`'s `signIn()` method that doesn't depend on `auth`, you would pass session details via the session provider or as explicit props in server components if using NextJS. The problem is just the middleware thing for which there appears to be an approach.

As a point of improvement on that @balazsorban44 I think it would be great to have more clarity on what exactly needs to be kept in the auth config for it to work properly vs what can be omitted. Email provider is an issue for this one, but callbacks and adapters can also have code that doesn't work on middlewares, unclear what is safe to remove and what isn't. Thanks!

**Comment by MoOx** [comment]
It's definitely a misconfig. But...

For example I want to create [a custom sign-in page](https://authjs.dev/guides/pages/signin).
According to doc, I made a `providerMap`. But when I import this map, I get an issue

```
./node_modules/nodemailer/lib/dkim/index.js:10:1
Module not found: Can't resolve 'fs'
```

Now that I have dig enough, I understand why I have this issue.
But it's totally not intuitive, especially if server/client components are unclear to you. And I think to this day, it's unclear for a lot of people.

I am getting this issue because my page `"use client"` for a given library and I don't really understand why I have to do it (actually now I do).

I think the error I got can happen to a lot of new comers, and should require some "trigger warning" doc or some more clear error (I know this part might be really complicated).

**Comment by Mr-kumar** [comment]
Module not found: Can't resolve 'nodemailer'

./src/actions/user.ts (5:1)

Module not found: Can't resolve 'nodemailer'
  3 | import { client } from "@/lib/prisma";
  4 | import { currentUser } from "@clerk/nextjs/server";
> 5 | import nodemailer from "nodemailer";
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  6 | import Stripe from "stripe";
  7 |
  8 | const stripe = new Stripe(process.env.STRIPE_CLIENT_SECRET as string);


can you tell why i get this error after installing all the thigns
